src/Estimators/MagnetizationDensity.h

```diff
@@ -0,0 +1,220 @@
+  void registerOperatorEstimator(hdf_archive& file) override;
+
+  /**
+  * This is a convenience function that handles \int_0^{2\pi} dx f(x).
+  * There are two provided methods for integrating this, so this function picks Simpson's
+  * or Monte Carlo to perform this integration, while keeping awareness of the integration
+  * interval and number of samples.
+  *
+  * @tparam Value type.  Real or complex in double or single precision.  Return type is same as fgrid type.
+  * @param[in] fgrid f(x), the function to integrate.  Assumed to be on a [0-2pi] interval, and if the grid isn't
+  *        random, it's assumed to be uniform.
+  * @return Value of integral.
+  */
+  template<class VAL>
+  VAL integrateMagnetizationDensity(const std::vector<VAL>& fgrid)
+  {
+    Real start  = 0.0;
+    Real stop   = 2 * M_PI;
+    Real deltax = (stop - start) / (nsamples_ - 1.0);
+    VAL val     = 0.0;
+    switch (integrator_)
+    {
+    case Integrator::SIMPSONS:
+      //There's a normalizing 2pi factor in this integral.  Take care of it here.
+      val = integrateBySimpsonsRule(fgrid, deltax) / Real(2.0 * M_PI);
+
+      break;
+    case Integrator::MONTECARLO:
+      //Integrand has form that can be monte carlo sampled.  This means the 2*PI
+      //is taken care of if the integrand is uniformly sampled between [0,2PI),
+      //which it is.  The following is just an average.
+      val = std::accumulate(fgrid.begin(), fgrid.end(), VAL(0));
+      val /= Real(nsamples_);
+      break;
+    }
+
+    return val;
+  };
+
+private:
+  MagnetizationDensity(const MagnetizationDensity& magdens) = default;
+
+
+  /**
+  * Generates the spin integrand \Psi(s')/Psi(s)* \langle s | \vec{\sigma} | s'\rangle for a specific
+  *  electron iat.  Since this is a vectorial quantity, this function returns sx, sy, and sz in their own
+  *  arrays.  
+  *
+  * @param[in] pset ParticleSet  
+  * @param[in] wfn TrialWaveFunction
+  * @param[in] iat electron index
+  * @param[out] sx x component of spin integrand
+  * @param[out] sy y component of spin integrand
+  * @param[out] sz z component of spin integrand
+  */
+  void generateSpinIntegrand(ParticleSet& pset,
+                             TrialWaveFunction& wfn,
+                             const int iat,
+                             std::vector<Value>& sx,
+                             std::vector<Value>& sy,
+                             std::vector<Value>& sz);
+
+  /**
+  *  Implementation of Simpson's 1/3 rule to integrate a function on a uniform grid. 
+  *
+  * @tparam Value type.  Real or complex in double or single precision.  Return type is same as fgrid type.
+  * @param[in] fgrid f(x), the function to integrate. 
+  * @param[in] gridDx, the grid spacing for the uniform grid.  Assumed to be consistent with size of fgrid.
+  * @return Value of integral.
+  */
+  template<class VAL>
+  VAL integrateBySimpsonsRule(const std::vector<VAL>& fgrid, Real gridDx) const
```

Comment:
see comment above

Reply:
No longer templated, but templated random grid generation is now explicitly implemented and instantiated in cpp.